// node-bebop/examples/GoDrone.js
"use strict";

// REGISTER PROTO-TYPES

Date.prototype.format=function(e){var t="";var n=Date.replaceChars;for(var r=0;r<e.length;r++){var i=e.charAt(r);if(r-1>=0&&e.charAt(r-1)=="\\"){t+=i}else if(n[i]){t+=n[i].call(this)}else if(i!="\\"){t+=i}}return t};Date.replaceChars={shortMonths:["Jan","Feb","Mar","Apr","May","Jun","Jul","Aug","Sep","Oct","Nov","Dec"],longMonths:["January","February","March","April","May","June","July","August","September","October","November","December"],shortDays:["Sun","Mon","Tue","Wed","Thu","Fri","Sat"],longDays:["Sunday","Monday","Tuesday","Wednesday","Thursday","Friday","Saturday"],d:function(){return(this.getDate()<10?"0":"")+this.getDate()},D:function(){return Date.replaceChars.shortDays[this.getDay()]},j:function(){return this.getDate()},l:function(){return Date.replaceChars.longDays[this.getDay()]},N:function(){return this.getDay()+1},S:function(){return this.getDate()%10==1&&this.getDate()!=11?"st":this.getDate()%10==2&&this.getDate()!=12?"nd":this.getDate()%10==3&&this.getDate()!=13?"rd":"th"},w:function(){return this.getDay()},z:function(){var e=new Date(this.getFullYear(),0,1);return Math.ceil((this-e)/864e5)},W:function(){var e=new Date(this.getFullYear(),0,1);return Math.ceil(((this-e)/864e5+e.getDay()+1)/7)},F:function(){return Date.replaceChars.longMonths[this.getMonth()]},m:function(){return(this.getMonth()<9?"0":"")+(this.getMonth()+1)},M:function(){return Date.replaceChars.shortMonths[this.getMonth()]},n:function(){return this.getMonth()+1},t:function(){var e=new Date;return(new Date(e.getFullYear(),e.getMonth(),0)).getDate()},L:function(){var e=this.getFullYear();return e%400==0||e%100!=0&&e%4==0},o:function(){var e=new Date(this.valueOf());e.setDate(e.getDate()-(this.getDay()+6)%7+3);return e.getFullYear()},Y:function(){return this.getFullYear()},y:function(){return(""+this.getFullYear()).substr(2)},a:function(){return this.getHours()<12?"am":"pm"},A:function(){return this.getHours()<12?"AM":"PM"},B:function(){return Math.floor(((this.getUTCHours()+1)%24+this.getUTCMinutes()/60+this.getUTCSeconds()/3600)*1e3/24)},g:function(){return this.getHours()%12||12},G:function(){return this.getHours()},h:function(){return((this.getHours()%12||12)<10?"0":"")+(this.getHours()%12||12)},H:function(){return(this.getHours()<10?"0":"")+this.getHours()},i:function(){return(this.getMinutes()<10?"0":"")+this.getMinutes()},s:function(){return(this.getSeconds()<10?"0":"")+this.getSeconds()},u:function(){var e=this.getMilliseconds();return(e<10?"00":e<100?"0":"")+e},e:function(){return"Not Yet Supported"},I:function(){var e=null;for(var t=0;t<12;++t){var n=new Date(this.getFullYear(),t,1);var r=n.getTimezoneOffset();if(e===null)e=r;else if(r<e){e=r;break}else if(r>e)break}return this.getTimezoneOffset()==e|0},O:function(){return(-this.getTimezoneOffset()<0?"-":"+")+(Math.abs(this.getTimezoneOffset()/60)<10?"0":"")+Math.abs(this.getTimezoneOffset()/60)+"00"},P:function(){return(-this.getTimezoneOffset()<0?"-":"+")+(Math.abs(this.getTimezoneOffset()/60)<10?"0":"")+Math.abs(this.getTimezoneOffset()/60)+":00"},T:function(){var e=this.getMonth();this.setMonth(0);var t=this.toTimeString().replace(/^.+ \(?([^\)]+)\)?$/,"$1");this.setMonth(e);return t},Z:function(){return-this.getTimezoneOffset()*60},c:function(){return this.format("Y-m-d\\TH:i:sP")},r:function(){return this.toString()},U:function(){return this.getTime()/1e3}}
// taken from http://jacwright.com/projects/javascript/date_format/


// INIT
var bebop = require("../.");
var drone = null;

function sayHowzit() {
	try {
		iLog("        _  _   _____                                _   _     _             
		iLog("	    _| || |_|  __ \             /\               | | | |   (_)            ");
		iLog("	   |_  __  _| |  | | _____   __/  \   _ __  _   _| |_| |__  _ _ __   __ _ ");
		iLog("	    _| || |_| |  | |/ _ \ \ / / /\ \ | '_ \| | | | __| '_ \| | '_ \ / _` |");
		iLog("	   |_  __  _| |__| |  __/\ V / ____ \| | | | |_| | |_| | | | | | | | (_| |");
		iLog("	     |_||_| |_____/ \___| \_/_/    \_\_| |_|\__, |\__|_| |_|_|_| |_|\__, |");
		iLog("	                                             __/ |                   __/ |");
		iLog("	                                            |___/                   |___/ ");
	
	}
	return true;
}
exports.sayHowzit = sayHowzit;

function doDrone() {
try {
iLog("Starting Drone Boot Sequence");

drone = bebop.createClient();
iLog("Drone Client Created");

iLog("Attempting to connect to Drone");
drone.connect(function() {
iLog("Connected with Drone");
initDroneEvents();
initDroneFlightPath();
iLog("Drone Flight Path Initiated");
});
} catch (consume) {
iLog("error in doDrone() " + consume);
landDrone();
}
return true;
}
exports.doDrone = doDrone;

// METHODS
async function initDroneFlightPath() {

try {
drone.takeOff();
await pauseExecution("take-off", 4000);
drone.stop();
drone.down(10);
await pauseExecution("lowering", 2000);
drone.stop();
await pauseExecution("stop", 1000);
/*
 * drone.forward(20); await pauseExecution("forward", 1000); drone.stop(); await
 * pauseExecution("stop", 1000); drone.backward(20); await
 * pauseExecution("backward", 1000); drone.stop(); await pauseExecution("stop",
 * 1000);
 */
// drone.right(10);
// drone.clockwise(10);
// await pauseExecution("clockwise", 2000);
// drone.stop();
// await pauseExecution("stop", 1000);
drone.counterClockwise(30);
await pauseExecution("counterClockwise", 2000);
drone.stop();
// await pauseExecution("stop", 1000);
landDrone();
drone.land();
await pauseExecution("landing", 1000);
iLog("Drone done");
process.exit(1);
} catch(consume) {
iLog("something crashed");
drone.stop();
drone.land();
iLog(consume);
}
}

function landDrone() {
try{
drone.stop();
drone.land();
} catch(consume) {
iLog("Error in landing drone");
drone.stop();
drone.land();
iLog(consume);
}
}
exports.landDrone = landDrone;

function initDroneEvents(){
 drone.on("battery", function (data) {
    iLog("Drone Battery Level: " + data);
  });
}

function pauseExecution(vLog, timeMS){

iLog(" >> Drone Data: " +  vLog);
return new Promise(resolve => setTimeout(resolve, timeMS));
}

function iLog(vMsg) {
try{
console.log(new Date().format('Ymd His') + ": " + vMsg);
} catch(consume) {
console.log("error in iLog land --> " + consume);
}
}

function sleep(ms) {
  return new Promise(resolve => setTimeout(resolve, ms));
}
